File: app/src/main/java/com/druidpyrcel/biotech/finaldilution/StartupActivity.java
package com.druidpyrcel.biotech.finaldilution;

import android.app.AlertDialog;
import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.View;
import android.view.WindowManager;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ListView;

import com.druidpyrcel.biotech.finaldilution.model.Solution;
import com.druidpyrcel.biotech.finaldilution.sqlite.DataProvider;

import java.util.List;

public class StartupActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_startup);
        final ApplicationContext appState = ((ApplicationContext) getApplicationContext());
        appState.setDb(new DataProvider(appState));

        Button newSolutionButton = (Button) findViewById(R.id.addNewSolutionButton);
        newSolutionButton.setOnClickListener(new OnNewSolutionButtonClickListener());

        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
    }

    @Override
    protected void onResume() {
        super.onResume();
        refreshSolutionList();
    }

    private void refreshSolutionList() {

        final ApplicationContext appState = ((ApplicationContext) getApplicationContext());
        List<Solution> solutionList = appState.getDb().getAllSolutions();
        if (!solutionList.isEmpty()) {
            //TODO Save and set last solution
            appState.setCurrentSolution(appState.getDb().getAllSolutions().get(0));
        }
        ListView solutionListView = (ListView) findViewById(R.id.solutionListView);
        ArrayAdapter<Solution> solutionListAdapter = new ArrayAdapter<>(
                this, android.R.layout.simple_list_item_1, solutionList);
        solutionListView.setAdapter(solutionListAdapter);
        solutionListView.setOnItemClickListener(new SolutionChooseListener());
    }

    private class OnNewSolutionButtonClickListener implements View.OnClickListener {

        @Override
        public void onClick(View v) {
            final EditText solutionNamePicker = new EditText(StartupActivity.this);
            final ApplicationContext appState = ((ApplicationContext) getApplicationContext());

            AlertDialog.Builder alertDialogBuilder = new AlertDialog.Builder(StartupActivity.this);
            alertDialogBuilder.setView(solutionNamePicker)
                    .setMessage("Enter new solution name: ")
                    .setCancelable(false)
                    .setPositiveButton("OK", new DialogInterface.OnClickListener() {
                        @Override
                        public void onClick(DialogInterface dialog, int which) {
                            if (solutionNamePicker.getText().length() != 0) {
                                Solution solution = new Solution();
                                solution.setName(solutionNamePicker.getText().toString());
                                appState.getDb().addSolution(solution);
                                refreshSolutionList();
                                appState.setCurrentSolution(appState.getDb().getSolution(solutionNamePicker.getText().toString()));
                            }
                        }
                    })
                    .setNegativeButton("Cancel", null);
            final AlertDialog alertDialog = alertDialogBuilder.create();

            solutionNamePicker.setOnFocusChangeListener(new View.OnFocusChangeListener() {
                @Override
                public void onFocusChange(View v, boolean hasFocus) {
                    if (hasFocus) {
                        alertDialog.getWindow().setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_STATE_ALWAYS_VISIBLE);
                    }
                }
            });
            alertDialog.show();
        }
    }

    private class SolutionChooseListener implements AdapterView.OnItemClickListener {

        @Override
        public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
            final ApplicationContext appState = ((ApplicationContext) getApplicationContext());
            appState.setCurrentSolution((Solution) parent.getAdapter().getItem(position));
            Intent intent = new Intent(StartupActivity.this, MainActivity.class);
            startActivity(intent);
        }
    }
}

File: app/src/main/java/com/labwhisper/biotech/finaldilution/solution/view/StartupActivity.kt
package com.labwhisper.biotech.finaldilution.solution.view

import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.ContextMenu
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.ListView
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.Observer
import com.labwhisper.biotech.finaldilution.ApplicationContext
import com.labwhisper.biotech.finaldilution.R
import com.labwhisper.biotech.finaldilution.genericitem.putExtraAnItem
import com.labwhisper.biotech.finaldilution.solution.CareTaker
import com.labwhisper.biotech.finaldilution.solution.Solution
import com.labwhisper.biotech.finaldilution.solution.appmodel.StartupAppModel
import com.labwhisper.biotech.finaldilution.util.button
import com.labwhisper.biotech.finaldilution.util.listView
import com.labwhisper.biotech.finaldilution.util.textView

class StartupActivity : AppCompatActivity() {

    private lateinit var appModel: StartupAppModel

    override fun onCreate(savedInstanceState: Bundle?) {
        appModel = StartupAppModel((applicationContext as ApplicationContext).solutionGateway)
        super.onCreate(savedInstanceState)
        Log.d(TAG, "On create Startup Activity")
        setContentView(R.layout.startup)
        val newSolutionButton = button(R.id.addNewSolutionButton)
        newSolutionButton.setOnClickListener {
            SolutionNameDialog(this, appModel).forCreate().show()
        }
        appModel.solutionList.observe(this,
            Observer { solutions -> solutions?.let { refreshSolutionList(solutions) } })
    }

    override fun onResume() {
        super.onResume()
        //TODO Remove after adding observer pattern to business model
        appModel.refresh()
    }

    private fun refreshSolutionList(solutionList: List<Solution>) {

        val solutionListView = listView(R.id.solutionListView)
        val solutionListAdapter = object : ArrayAdapter<Solution>(
            this,
            R.layout.solution_list_item,
            R.id.solution_list_text1,
            solutionList
        ) {
            override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
                val view = super.getView(position, convertView, parent)
                val solution = solutionList[position]
                view.textView(R.id.solution_list_text1).text = solution.name
                view.textView(R.id.solution_list_text2).text = solution.displayString()
                return view
            }
        }
        solutionListView.adapter = solutionListAdapter
        solutionListView.setOnItemClickListener { parent, _, position, _ ->
            enterSolution(parent.adapter.getItem(position) as Solution)
        }
        registerForContextMenu(solutionListView)
    }

    override fun onCreateContextMenu(
        menu: ContextMenu?, v: View?, menuInfo: ContextMenu.ContextMenuInfo?
    ) = menuInflater.inflate(R.menu.menu_solution, menu)

    override fun onContextItemSelected(item: MenuItem?): Boolean {
        if (item == null) {
            return super.onContextItemSelected(null)
        }
        return when (item.itemId) {
            R.id.action_rename -> {
                val info = item.menuInfo as AdapterView.AdapterContextMenuInfo
                val listView = info.targetView.parent as ListView
                val solution = listView.adapter.getItem(info.position) as Solution
                SolutionNameDialog(this, appModel).forRename(solution).show(); true
            }
            R.id.action_clone -> {
                val info = item.menuInfo as AdapterView.AdapterContextMenuInfo
                val listView = info.targetView.parent as ListView
                val solution = listView.adapter.getItem(info.position) as Solution
                SolutionNameDialog(this, appModel).forClone(solution).show(); true
            }
            R.id.action_delete -> {
                val info = item.menuInfo as AdapterView.AdapterContextMenuInfo
                val listView = info.targetView.parent as ListView
                val solution = listView.adapter.getItem(info.position) as Solution
                DeleteConfirmDialog(this, appModel).create(solution).show(); true
            }
            else -> false
        }
    }

    private fun enterSolution(solution: Solution) {
        val intent = Intent(this@StartupActivity, EditActivity::class.java)
        intent.putExtraAnItem(solution)
        intent.putExtra("CARE_TAKER", CareTaker<Solution>())
        startActivity(intent)
    }

    private fun deleteSolution(solution: Solution) {
        appModel.deleteSolution(solution)
    }

    companion object {
        private val TAG = "Startup Activity"
    }

}
